ユーザーアイコン機能追加

ItohYuto opened the pull request:
- ユーザー編集画面にて追加、削除可能
- プレビュー機能あり
- N+1対策

ItohYuto committed:
ユーザーアイコン機能追加

maedana commented on Gemfile:
```diff
@@ -48,7 +48,7 @@ gem 'bootsnap', require: false
 
 # Use Active Storage variants [https://guides.rubyonrails.org/active_storage_overview.html#transforming-images]
-# gem "image_processing", "~> 1.2"
+gem 'image_processing', '~> 1.2'
```
元からなので修正は任意で結構ですが、Railsのように勝手にバージョンが上がって欲しくない一部のgemを除き、原則はバージョン指定なしで最新版を使えばよいです。

ItohYuto replied:
修正しました。

maedana commented on app/javascript/registrations.js:
```diff
@@ -0,0 +1,7 @@
+document.getElementById('user_image').onchange = function() {
```
Railsのプラクティスですしまあり目くじら立てなくていいかなと思いつつ気になったので書いておきます。

最近はgetElementByIdをあまり見かけないイメージがあります。大抵以下を使ってることが多い気がします。
https://developer.mozilla.org/ja/docs/Web/API/Document/querySelector
https://developer.mozilla.org/ja/docs/Web/API/Document/querySelectorAll

あとfunctionも特別な理由がなければアロー関数使うことが多いかなーと。

ItohYuto replied:
getElementBy\~をquerySelectorに、functionをアロー関数に修正しました。
確認なのですが、querySelectorの方がよくつかわれている理由としてはコーディングにおける汎用性が良いからでしょうか？
取得する要素についてはgetElementsByClassNameはHTMLCollection、querySelectorAllはNodeListと違いがあったりするのに加えて、処理速度としてもquerySelectorとgetElementBy\~ではGetElementByの方が速いとされています。
特にHTMLが長いわけではないかつ多くの処理があるわけでもないので速度は今回不要かと思いますが、それでもなぜquerySelectorの方がつかわれているのかが気になりました。

個人的にはgetElementBy~の方が名前的にもわかりやすいなと感じていて、コーディングにおける汎用性以外にquerySelectorを使用する利点が分かっておらず教えていただきたいです。

maedana replied:
フロントエンド系あまり得意ではなく、必要に迫られてJavaSciprt書いてるだけですけど`querySelector`ではCSSセレクタで指定できるので、ある程度CSSセレクタになれていると便利かなシーンが多いかなとは思います。処理速度の面の話はそうかもしれませんが、そこが問題になるほどのものって昨今はそうそうないのではないですかね。

とはいえ、https://qiita.com/ari-chel/items/b06c68aec8849d0409dd 等参照してみるとそれなりに巨大なDOMを操作するときは無視出来ないような差も出そうですね。両方を意識しつつ普段は`querySelector`がお手軽なのかな、と感じました。

たぶんご納得のいく回答ではないと思いますので、気になるようであればFBCのQ&Aで聞いてみると有識者からの回答もあつまるような気がします。

maedana commented on app/javascript/registrations.js:
```diff
@@ -0,0 +1,7 @@
+document.getElementById('user_image').onchange = function() {
+  var fileReader = new FileReader();
```
varではなくconstにしたほうがよさそうですね。

ItohYuto replied:
修正しました。

maedana commented on app/views/devise/registrations/_profile_fields.html.erb:
```diff
@@ -1,3 +1,23 @@
+<div class="field">
+  <%= f.label :image %><br />
+  <%= f.file_field :image %><br />
+  <div id="preview_box">
```
未使用なidですかね? であればつけなくていいかなと思いました。

ItohYuto replied:
不要なidのため削除しました。

maedana commented on app/views/devise/registrations/_profile_fields.html.erb:
```diff
@@ -1,3 +1,23 @@
+  <%= f.label :image %><br />
+  <%= f.file_field :image %><br />
+  <div id="preview_box">
+    <img id="preview" width="100">
```
このままでよいのですが、ご参考までに書いておきます。

プレビュー用のjsで利用してるからつけているidだと理解しました。このようなケースで、1つしかない場合でも最近は割とclass使うイメージがあるかもです。あとjs利用でデザイン用でないとすぐわかるように弊社ですと、`class="js-preview"`のようにprefixをつけたりもします。そのうえで、`querySelector`使ってDOM取得する感じですかね。

ItohYuto replied:
jsで利用する際はidよりclass指定のが多いのは言われてHTML構造と関係ない名前をidに付けるのは違うなぁと感じました。
言語化まではうまくできないですが、なんとなく想像できるような気がします。
prefixに関しては利用先がわかりやすいので、今後意識していきたいと思います。

maedana commented on app/controllers/users/registrations_controller.rb:
```diff
@@ -0,0 +1,64 @@
+  # PUT /resource
+  def update
+    super
+    resource.image.attach(account_update_params[:avatar]) if account_update_params[:avatar].present?
```
imageという名前のパラメタで飛んできてますし、
app/controllers/application_controller.rbで`configure_permitted_parameters`でstrong parameterの制御出来てますから、これは意味のないコードになってるのではないですかね

ItohYuto replied:
おっしゃる通り不要な処理でしたので、修正しました。

maedana commented on app/controllers/users/registrations_controller.rb:
```diff
@@ -0,0 +1,64 @@
+  def update
+    super
+    resource.image.attach(account_update_params[:avatar]) if account_update_params[:avatar].present?
+    @user.image.purge if params[:image_delete]
```
ちょっとトリッキーなやり方かなと思いました。単に削除したいだけなら画像削除用のcontroller及びactionを用意してdeviseのカスタマイズしない方がシンプルかなーと思いました。画面もチェックボックスつけて更新というよりはシンプルな削除ボタンや、削除リンクのイメージですね。なお、動作上問題なさそうですし要修正とはまでは思ってないです。

ItohYuto replied:
|メソッド|URL|処理|
|:----|:----|:----|
|DELETE|/user/user_id/attahments|ユーザー情報のアタッチメントを削除する。|

上記のような形で、実装しました。
今回は画像のみでユーザーと1対1のため、attachments_idは不要なURL設計としました。
フォーム内に別でボタンを追加できないため、削除リンクを作成しました。

ItohYuto committed:
指摘箇所修正

- Gemの不要なバージョン指定削除
- Js:無名関数をアロー関数で記述するように変更。
- Js:変更を加えない変数をconstで定義
- Js:要素取得時はgetElementBy~ではなく。querySelectorを使用するように変更。
- html:不要なid削除
- html:jsで利用するためのidはクラスに変更し、名前に'js-'のprefixを付与。
- 不要な記述を削除
- 画像削除用のリンクを作成。

ItohYuto commented:
レビューありがとうございます。
修正しましたので再度レビューお願いいたします。

## スクリーンショット
- アカウント編集ページ(画像なし・プレビューなし)
![image](https://github.com/ItohYuto/fjord-books_app-2023/assets/88817643/91331de4-2bd6-45fc-a67f-f9650e70eefc)
- アカウント編集ページ(画像なし・プレビューあり)
![image](https://github.com/ItohYuto/fjord-books_app-2023/assets/88817643/b84639d0-cf62-45ea-91e1-5b7e79bdf065)
- アカウント編集ページ(画像あり)
![image](https://github.com/ItohYuto/fjord-books_app-2023/assets/88817643/bd82d2ba-1dc5-48fe-93ca-50fbd5af11de)
- ユーザ一覧ページ
![image](https://github.com/ItohYuto/fjord-books_app-2023/assets/88817643/d9ec9b69-f96d-4c09-8924-f075c297a4ab)
- ユーザー詳細ページ
![image](https://github.com/ItohYuto/fjord-books_app-2023/assets/88817643/7b7c39ff-c384-4120-be17-c24ce8b6f952)
## rubocop-fjord実行結果
![image](https://github.com/ItohYuto/fjord-books_app-2023/assets/88817643/43e41c10-8b63-4d94-bec3-419f752a2b66)
## erblint実行結果
![image](https://github.com/ItohYuto/fjord-books_app-2023/assets/88817643/a4381ff6-6897-4c45-ab43-8826bac5a050)
